### ISSUE TITLE
Talos does float badly

### ISSUE THREAD

**Issue by renxida** [autonomio/talos#135]
Thanks so much for coming here to raise an issue. Please take a moment to 'check' the below boxes:

- [ * ] I'm up-to-date with the latest release:
    
      pip install -U talos

- [ * ] I've confirmed that my Keras model works outside of Talos.

If you still have an error, please submit **complete trace** and a code with: 

- output of shape for x and y e.g. (212,12)
- Talos params dictionary
- The Keras model wired for Talos
- Description of extra variables in the model

You can provide the code in pastebin / gist or any other format you like. 

-------------------------------------------------------------------------

**Comment by renxida** [comment]
I hereby confirm that my Talos is up to date, and that my model function works outside Talos : 2

**Comment by renxida** [comment]
Model function:
```
from keras.models import Model
from keras.layers import Input, Dense, Conv1D, Flatten, Concatenate
from keras import regularizers

def make_net(input_layer, params):
    net = input_layer

    if params.get('conv_count', 0) !=0:
        for _ in range(params['conv_count']):
            newConv = Conv1D(
                    activation='relu',
                    filters=params['conv_filters'],
                    kernel_size=params['conv_kernel_size'],
                    kernel_regularizer=regularizers.l2(float(params['l2reg'])))
            net = newConv(net)
            # optional pooling
            if params.get('pool_size', None) is not None:
                newPool = keras.layers.MaxPooling1D(
                        pool_size=params['pool_size'],
                        strides=params.get(
                            'pool_strides',
                            params['pool_size']),
                        )
                net = newPool(net)

        net = Flatten()(net)
        
    if params.get('dense_count', 0) !=0:
        for dense_num in range(params['dense_count']):
            newDense = Dense(
                    units=params['dense_units'],
                    activation='relu',
                    kernel_regularizer=regularizers.l2(float(params['l2reg'])),
                    )
            net = newDense(net)

    return net

def make_optimizer(params):
    opzr = keras.optimizers.Adam(**params.get('optimizer', {}))
    return opzr

import keras.backend as K
def model(params):
    
    
    def metric_correlation(y_true, y_pred):
        logger.warn(f'[metric_correlation] y_true size: {y_true.shape}, y_predy size: {y_pred.shape}')

        def m(x, w):
            """Weighted Mean"""
            return K.sum(x*w) / K.sum(w)
        def cov(x, y, w):
            """Weighted Covariance"""
            return K.sum(w* (x - m(x, w))*(y - m(y, w)))/ K.sum(w)

        def corr(x, y, w):
            """Weighted Correlation"""
            return cov(x, y, w) / K.sqrt(cov(x, x, w) * cov(y, y, w))
        print(y_true.get_shape())
        return corr(y_true, y_pred, K.ones_like(y_true))
    inputs = []

    subnets = []


    inpt = Input(shape=(30,), dtype='float32', name='returns_input')
    inputs.append(inpt)

    net = make_net(inpt, params)
    
    condenser = Dense(units=1, activation=None, use_bias=False)
    net = condenser(net)
    
    model = Model(inputs=inputs, outputs = net)

    opzr = make_optimizer(params)
    model.compile(
        loss='mean_squared_error',
        optimizer=opzr,
        metrics=[metric_correlation])
    return model
```

**Comment by renxida** [comment]
To reproduce problem, run with:
```
h = ta.Scan(x=np.random.normal(loc=0.0, scale=1.0, size=(5000, 30)),
            y=np.random.normal(loc=0.0, scale=1.0, size=(5000, 1)), 
            params={
                'dense_count': [1, 2, 3],
                'dense_units': [50, 100, 150, 200],
                'conv_count': [0],
                'l2reg': [1.0, 100, 1000],
                'epochs':[4],
                'batch_size':[32]
            },
            model=model_ta,
            grid_downsample=0.1,
            functional_model=True)
```

**Comment by renxida** [comment]
PS: I know the solution to this problem is to do float-to-int conversion for each parameter supplied to `model_ta()`, but I would like to have talos and not me handle this.

**Comment by renxida** [comment]
PPS: I just pulled the newest commit on daily-dev 

> @mikkokotila
mikkokotila Merge pull request #124 from awilliamson/early_stopper

and the problem persists.

**Comment by mikkokotila** [comment]
Regarding the error you are getting, did I gather it correctly and you are able to run the experiment when all the values are ints?

Unfortunately we can't change all floats to ints, as some parameters are floats (like dropout and lr). There could of course be some simple system that converts values to ints if all values are whole numbers in float form. That way there is no possible downside it seems. 

That said, the error you are getting might have to do with something other than what it seems to have to do. As I see, the error itself comes from trying to use a float for range (or so it seems). I have witnessed rare cases where something similar happened with a similar effect, where actually the error messages gave zero indication of the actual issue (?!). These were caused by:

### 1) model is not in standard format

First regarding the input model, to be safe, the format always has to be:

```
def model_ta(x_train, y_train, x_val, y_val, params):

    # the model contents is here
    
    return out, model

```
The key point here is that the model arguments has to be exactly like that, and the return has to be exactly like that as well (where model that is created inside the function, and out is the history object),. 

2) some of the parameters that are expected to follow standard naming convention (eg. 'first_neuron' and 'dropout') were using some other names 

In such cases (these were never reported but is my own experience arising from extensive use) 0 of the time the error saying something about dtype had something to do with dtype. If this happened 10 times, then 9 times it was because of one of the two reasons above. 

Can you check these in your end and see what happens then.

**Comment by renxida** [comment]
- model_format: I'm dumb. Forgot to include this: (I'll give you a full quote that actually replicates everything)

```
def model_ta(x_train, y_train, x_val, y_val, params):
    print(params)
    # replace the hyperparameter inputs with references to params dictionary 
    mdl = model(params)
    
    history_callback = mdl.fit(x_train, y_train,
                batch_size=params['batch_size'],
                epochs=params['epochs'],
                verbose=1,
                validation_data=[x_val, y_val])
    history_callback.history['val_acc']=[42]
    # modify the output model
    return history_callback, mdl
```

- dtype issue: I'm pretty sure the issue is caused by talos though. the thing is this:
   when I provide a single float for l2reg, all the OTHER parameters get changed to float. So things like layer count and number of neurons in the layers. Parameters should have their dtypes handled individually, amiright?

- naming convention: thanks for telling me that. I'm new to ML so I know nothing.

I'll compile a file with the appropriate code to replicate the problem; expect it in a few minutes.

**Comment by renxida** [comment]
https://gist.github.com/renxida/0f708665aeff7bca2fdbbbae9a147462

**Comment by renxida** [comment]
this is the first time in my lifetime that I got an issue with a label like 'high priority fix' : )

Thanks. It makes me feel incredibly happy about contributing to this project.

**Comment by mikkokotila** [comment]
Nice. Thanks :)

It seems that you found something interesting there with forced float conversion, this should clearly not happen. I will go through the codes to identify where and how this happens. I changed the name title of issue to reflect our current understanding of it.

**Comment by mikkokotila** [comment]
When numpy array is created from a list, if the list has ints and floats, it's all converted to floats. If there is also string (and floats and ints) nothing is converted. 

So any case where only int and float is present in params dictionary, and as long as there is even a single float, all the values will be converted to float. That's clearly a problem. The issue is created here: 

 ```
   def _param_grid(self):

        '''CREATE THE PARAMETER PERMUTATIONS

        This is done once before starting the experiment.
        Takes in the parameter dictionary, and returns
        every possible permutation in an array.
        '''

        ls = [list(self._p[key]) for key in self._p.keys()]
        _param_grid_out = array(list(product(*ls)))

        return _param_grid_out
```
The fix is to declare `dtype='object'` in the array creation. It's now merged up until dev. The right version is 0.4.5. 

Thanks a lot for helping, it's really great to have this bug out.

**Comment by renxida** [comment]
Glad I could help.

Over the summer of 2018 I've been doing some hyperparameter searching with a set of scripts I've written. They work, but they're a mess, and I'm looking to migrate everything to Talos, which is superior in that it's better organized.

Expect more issues & PRs from me over the winter break as I perform the migration : )
